diff --git a/ui/sdl2-gl.c b/ui/sdl2-gl.c
index 28d796607c..48f41a59f7 100644
--- a/ui/sdl2-gl.c
+++ b/ui/sdl2-gl.c
@@ -97,7 +97,12 @@ void sdl2_gl_switch(DisplayChangeListener *dcl,
     }
 
     if (!scon->real_window) {
-        sdl2_window_create(scon);
+      sdl2_window_create(scon);
+      SDL_GL_MakeCurrent(scon->real_window, scon->winctx);
+      SDL_GL_SetAttribute(SDL_GL_SHARE_WITH_CURRENT_CONTEXT, 1);
+      SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE);
+      SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 4);
+      SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 1);
         scon->gls = qemu_gl_init_shader();
     } else if (old_surface &&
                ((surface_width(old_surface)  != surface_width(new_surface)) ||
diff --git a/ui/shader.c b/ui/shader.c
index ab448c41d4..9d980ee809 100644
--- a/ui/shader.c
+++ b/ui/shader.c
@@ -158,7 +158,7 @@ QemuGLShader *qemu_gl_init_shader(void)
     gls->texture_blit_flip_prog = qemu_gl_create_compile_link_program
         (texture_blit_flip_vert_src, texture_blit_frag_src);
     if (!gls->texture_blit_prog || !gls->texture_blit_flip_prog) {
-        exit(1);
+      fprintf(stderr, "qemu_gl_int_shader failed\n");
     }
 
     gls->texture_blit_vao =
diff --git a/ui/shader/texture-blit-flip.vert b/ui/shader/texture-blit-flip.vert
index f7a448d229..112320e5df 100644
--- a/ui/shader/texture-blit-flip.vert
+++ b/ui/shader/texture-blit-flip.vert
@@ -1,4 +1,4 @@
-#version 300 es
+#version 410 core;
 
 in vec2  in_position;
 out vec2 ex_tex_coord;
diff --git a/ui/shader/texture-blit.frag b/ui/shader/texture-blit.frag
index 8ed95a46b6..a1262dada2 100644
--- a/ui/shader/texture-blit.frag
+++ b/ui/shader/texture-blit.frag
@@ -1,4 +1,4 @@
-#version 300 es
+#version 410 core;
 
 uniform sampler2D image;
 in  mediump vec2 ex_tex_coord;
diff --git a/ui/shader/texture-blit.vert b/ui/shader/texture-blit.vert
index fb48d70665..b93125e6f7 100644
--- a/ui/shader/texture-blit.vert
+++ b/ui/shader/texture-blit.vert
@@ -1,4 +1,4 @@
-#version 300 es
+#version 410 core
 
 in vec2  in_position;
 out vec2 ex_tex_coord;
